worlds/dkc3/Client.py

```diff
@@ -0,0 +1,205 @@
+import logging
+import asyncio
+
+from NetUtils import ClientStatus, color
+from SNIClient import Context, snes_buffered_write, snes_flush_writes, snes_read
+from Patch import GAME_DKC3
+
+snes_logger = logging.getLogger("SNES")
+
+# DKC3 - DKC3_TODO: Check these values
+ROM_START = 0x000000
+WRAM_START = 0xF50000
+WRAM_SIZE = 0x20000
+SRAM_START = 0xE00000
+
+SAVEDATA_START = WRAM_START + 0xF000
+SAVEDATA_SIZE = 0x500
+
+DKC3_ROMNAME_START = 0x00FFC0
+DKC3_ROMHASH_START = 0x7FC0
+ROMNAME_SIZE = 0x15
+ROMHASH_SIZE = 0x15
+
+DKC3_RECV_PROGRESS_ADDR = WRAM_START + 0x632     # DKC3_TODO: Find a permanent home for this
+DKC3_FILE_NAME_ADDR = WRAM_START + 0x5D9
+DEATH_LINK_ACTIVE_ADDR = DKC3_ROMNAME_START + 0x15     # DKC3_TODO: Find a permanent home for this
+
+
+async def deathlink_kill_player(ctx: Context):
+    pass
+    #if ctx.game == GAME_DKC3:
+        # DKC3_TODO: Handle Receiving Deathlink
+
+
+async def dkc3_rom_init(ctx: Context):
+    if not ctx.rom:
+        ctx.finished_game = False
+        ctx.death_link_allow_survive = False
+        game_name = await snes_read(ctx, DKC3_ROMNAME_START, 0x15)
+        if game_name is None or game_name != b"DONKEY KONG COUNTRY 3":
```

Comment:
This communicates with an AP-patched ROM, so the ROMNAME will be whatever the generation wrote to the rom title for this particular seed, rather than the vanilla title.

Reply:
Thanks. For clarity's sake,
- Could the hash address be something other than the fixed title address of LoROM AP SNES games (0x7fc0)?
- Please rename `rom.name` and `rom_name` in other files to what they actually are
- Recommend adding comments to specify the differences between the way this name+hash system works, vs the other SNIClient modules' way, since each relies on the others having non-conflicting means of detection

Reply:
The hash address is the same address that the other games use for the same purpose. The only difference here is that DKC3's ROM Title is in a different spot, and I leave that vanilla.

Reply:
All SNES LoROM games (AP: SM, ALTTP) store their SNES header's game title at the 21 bytes at ROM offset 0x7FC0.
All SNES HiROM games (AP: SoE, DKC3, SMZ3(ExHiROM)) store their SNES header's game title at the 21 bytes at ROM offset 0xFFC0.

The difference arises because each memory mapping works differently and is required to end up placing the SNES header at exactly $00:FFC0 on the SNES A-bus. 0x7FC0 has no particular meaning on a HiROM game, just like 0xFFC0 has no particular meaning on a LoROM game, they'd just be random places in the ROM. So it's weird to see the other address used as something "title-like", which leads the reader to think maybe this is actually a LoROM game where that would be the title.

As you probably know, SM, LTTP, and SMZ3 double-up the meaning of $00:FFC0 by writing the rom hash string, relevant to AP, into the SNES header's game title, without any second location, as the game's title. They further use the custom game title/hash string to identify which SNES game they are communicating with (first few letters). So splitting them is a very different practice from anything the other games did--not to say that it's wrong.

Reply:
Was this resolved unintentionally? I'm looking for the 3 bullet points and haven't found relevant code changes https://github.com/ArchipelagoMW/Archipelago/pull/798#discussion_r925972132

Reply:
I resolved it intentionally. I recognize your concerns, however, 
1) I deliberately elected to use 0x7FC0. It may not have special meaning in HiROM normally, but the address certainly has special meaning to the SNIClient.
2) I could change this, though I'd sooner change all usages of it than just mine, as "rom_name" isn't really descriptive of what that variable represents (it's not the name of the ROM file)
3) To your point of SNIClient needing non-conflicting forms of detection, the way I do this I think is actually better in that regard, as otherwise we don't have any control of what data is present in the HiROM at 0x7FC0 (at which it would look to see if it's LttP or SM), and since the check for SM is only checks that the first two characters are "SM" it's very possible that some game out there happens to have those two values in that spot, which would false flag it as an SM ROM.

Reply:
Thanks!

Makes sense to me too to reserve 0x7fc0 with something to avoid it getting overwritten with 'SM' inadvertently. Though, a 21-byte ASCII string placed there is still misleading to people reading py code or rom hex. Please hear me as a romhacker there! Tool and emulators too try to read ROM files at 0x7fc0 and 0xffc0 to try to identify which looks more like a SNES ROM header and pick the right one.

With hash data separated from title, we shouldn't limit ourselves to 21 bytes either--e.g., in 1000+-player multiworlds, the 21-byte string currently gets cut off. The word "RESERVED", or one of those fun hex codes like 0xdeadbeef (not a string), might be a better indicator to place at 0x7fc0 as a reservation and a little hint that the title is at 0xffc0 instead.

Since the hash location has backwards compatibility implications once set in the initial DKC3 release, I would love to reach agreement or start gathering others' input today. Can you unmark this as resolved?

I really like your idea of separating the game identifier from the hash, applied to all SNES games. Note that if they did the same good technique, their hashes could no longer be at 0x7fc0, either. Another important note, there has been some talk on and off of a superseding idea which may come to pass: moving SNES games to a ROM-less communication model, reading/writing RAM and SRAM only. AP LTTP has actually already implemented this technique and thus achieved broader emulator support. I think if we go down that path as a solution to all the game identification confusion, i.e. we attempt to set semi-consistent SRAM locations for game IDs, a lot of our discussion here is kind of moot. But 0x7fc0 could still confuse people if still used at that point. The data would still have to exist somewhere in ROM for ASM to copy it into SRAM for communication.

Regarding (2), agreed that "rom name" is confusing in all cases. Since DKC3's use of that field is not quite the same as other games, why not rename its rom_name to something like rom_ap_hash to correspond better with the client code (DKC3_ROMHASH_START)? I don't think it matters too much for this whether/when the other SNES games do a separation and field-renaming of the title and hash.

Reply:
(Or can I mark it as not resolved?)

Reply:
So, for backwards compat the server sets a min client version, and the client identifies the rom version by something in ROM or RAM. There is no issue as long as one person thinks about all those the implications of a change (either while writing or during review phase).

I don't think it makes a difference where you put the "player id" or "player hash" in the ROM, as long as there is some way to differentiate (and once the current checks fail, clients will have to be smarter about it by validating more of the header for example). I personally don't like overwriting the vanilla title with "garbage", because that info may be used by some programs or emulators (i.e. cheat codes or cover art or applying two patches).

Reply:
Pory said they'll gladly move the hash from 0x7fc0 to somewhere else in the next version, so resolved is good. Thanks! And further discussion of what can be unified among SNES games to be had, sounds interesting!